ark/jsonschema/__tests__/object.test.ts

```diff
@@ -0,0 +1,148 @@
+	})
+
+	it("minProperties", () => {
+		const tMinProperties = parseJsonSchema({
+			type: "object",
+			minProperties: 2
+		})
+		attest(tMinProperties.json).snap({
+			domain: "object",
+			predicate: ["$ark.jsonSchemaObjectMinPropertiesValidator"]
+		})
+		attest(tMinProperties.allows({})).equals(false)
+		attest(tMinProperties.allows({ foo: 1 })).equals(false)
+		attest(tMinProperties.allows({ foo: 1, bar: 2 })).equals(true)
+		attest(tMinProperties.allows({ foo: 1, bar: 2, baz: 3 })).equals(true)
+	})
+
+	it("properties & required", () => {
+		const tRequired = parseJsonSchema({
+			type: "object",
+			properties: {
+				foo: { type: "string" },
+				bar: { type: "number" }
+			},
+			required: ["foo"]
+		})
+		attest(tRequired.json).snap({
+			domain: "object",
+			required: [{ key: "foo", value: "string" }],
+			optional: [{ key: "bar", value: "number" }]
+		})
+
+		attest(() => parseJsonSchema({ type: "object", required: ["foo"] })).throws(
+			"TraversalError: must be a valid object JSON Schema (was an object JSON Schema with 'required' array but no 'properties' object)"
+		)
+		attest(() =>
+			parseJsonSchema({
+				type: "object",
+				properties: { foo: { type: "string" } },
+				required: ["bar"]
+			})
+		).throws(
+			`TraversalError: required must be a key from the 'properties' object (one of ["foo"]) (was bar)`
+		)
+		attest(() =>
+			parseJsonSchema({
+				type: "object",
+				properties: { foo: { type: "string" } },
+				required: ["foo", "foo"]
+			})
+		).throws(
+			`TraversalError: required must be an array of unique strings (was an array with the following duplicates: [{"element":"foo","indices":[1]}])`
+		)
+	})
+
+	it("additionalProperties", () => {
+		const tAdditionalProperties = parseJsonSchema({
+			type: "object",
+			additionalProperties: { type: "number" }
+		})
+		attest(tAdditionalProperties.json).snap({
+			domain: "object",
+			predicate: ["$ark.jsonSchemaObjectAdditionalPropertiesValidator"]
+		})
+		attest(tAdditionalProperties.allows({})).equals(true)
+		attest(tAdditionalProperties.allows({ foo: 1 })).equals(true)
+		attest(tAdditionalProperties.allows({ foo: 1, bar: 2 })).equals(true)
+		attest(tAdditionalProperties.allows({ foo: 1, bar: "2" })).equals(false)
+	})
+
+	it("patternProperties", () => {
+		const tPatternProperties = parseJsonSchema({
+			type: "object",
+			patternProperties: {
+				"^[a-z]+$": { type: "string" }
+			}
+		})
+		attest(tPatternProperties.json).snap({
+			domain: "object",
+			index: [
+				{
+					signature: { domain: "string", pattern: ["^[a-z]+$"] },
+					value: "string"
+				}
+			]
+		})
+		attest(tPatternProperties.allows({})).equals(true)
+		attest(tPatternProperties.allows({ foo: "bar" })).equals(true)
+		attest(tPatternProperties.allows({ foo: 1 })).equals(false)
+		attest(tPatternProperties.allows({ "123": "bar" })).equals(true) // true since allows additional properties
+	})
+
+	it("propertyNames", () => {
```

Comment:
hopefully we can convert this implementation over to a built-in index signature and we don't need to test 🙏

Reply:
Yep! This was done across https://github.com/arktypeio/arktype/commit/78aea2ff4a2a6bb93557e34e21a5a94fde874a7a, https://github.com/arktypeio/arktype/commit/0fc22f46dad8c56c3d868b074689d5b4f068061e, https://github.com/arktypeio/arktype/commit/d61853eb9034a55cbbe05aecd58170322822f9fb, https://github.com/arktypeio/arktype/commit/3644bbf32cb4986a8f6a687245d779a179dfd543